### PR TITLE
change(doc): Document state format upgrade implementation and testing

### LINKED ISSUE #7737: doc: Document how to implement state format upgrades
### Motivation

We have learnt a lot about in-place state format upgrades over the past few months, let's write the most important parts down.

### Rough Plan

Background:
- [x] explain two different ways to sync, full and upgrade

Design:
- [x] upgrade
  - [x] read methods must be able to read both formats
  - [x] write methods should be refactored so they are called from both the full sync and upgrade code, this prevents data inconsistencies
- [x] quick and detailed format checks
  - always runs on startup, shutdown, and after upgrades
  - runs periodically in CI using a custom config

Development:
- [x] how to edit or reset the state version file during development for manual testing

### Review

We should get multiple developers to read and review this documentation, to make sure we all understand and agree with it.

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We want to document what we've learned from the last 3 in-place state upgrades.

Close #7737

## Solution

Document:
- how the state works
- block verification modes
- why state upgrades are needed
- why they are tricky and need special care
- how to implement them more reliably (see also the checklist link below)
- what testing is needed

Also update some outdated format docs, and add some headings.

There are also ticket templates coming in #7738, here is a draft implementation checklist for state upgrades:
https://github.com/ZcashFoundation/zebra/issues/7738#issuecomment-1774217120

## Review

We should get two reviews on this one.

### Reviewer Checklist

  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

Reference these docs from a state upgrade ticket template.

### CODE REVIEW

File: book/src/dev/state-db-upgrades.md [hunk 1]
```diff
@@ -1,18 +1,63 @@
...
 which can take days.
 
+### Writing Blocks to the State
+[write-block]: #write-block
+
+Zebra's state has two methods of getting to the latest format:
+- syncing from empty with the latest format, and
+- upgrading from a previous format to the latest format.
+
+This means that writing blocks uses two different codepaths, and they must produce the same results.
```
Comment (arya2) [inline book/src/dev/state-db-upgrades.md]: Optional/Nitpick: I think "both must produce the same results" should go in the first line of this section.

```suggestion
Blocks can be written to the database via two different code paths, and both must produce the same results:

- Upgrading pre-existing zebra-state databases/cache to the latest format
- Writing newly-synced blocks in the latest format

```

File: book/src/dev/state-db-upgrades.md [hunk 2]
```diff
@@ -1,18 +1,63 @@
...
+To reduce code and testing complexity:
+- when a previous Zebra version opens a newer state, the entire state is considered to have that lower version, and
+- when a newer Zebra version opens an older state, each required upgrade is run on the entire state.
+
 ### In-Place Upgrade Goals
+[upgrade-goals]: #upgrade-goals
 
+Here are the goals of in-place upgrades:
 - avoid a full download and rebuild of the state
 - the previous state format must be able to be loaded by the new state
```
Comment (arya2) [inline book/src/dev/state-db-upgrades.md]: Nitpick/Optional:

```suggestion
- Zebra's state must be able to upgrade data from the previous major version of its disk format
  and load data from the previous minor or patch versions of its disk format
```

File: book/src/dev/state-db-upgrades.md [hunk 3]
```diff
@@ -62,17 +116,50 @@ But we need to use iterators for some operations, so our alternatives are (in pr
...
+#### Manually Triggering a Format Upgrade
+[manual-upgrade]: #manual-upgrade
+
+Zebra stores the current state minor and patch versions in a `version` file in the database
+directory. This path varies based on the OS, major state version, network, and config.
+
+For example, the default mainnet state version on Linux is at:
+`~/.cache/zebra/state/v25/mainnet/version`
+
+To upgrade from `v25.0.0`, delete the version file.
```
Comment (arya2) [inline book/src/dev/state-db-upgrades.md]: Nitpick/Optional:

```suggestion
To upgrade a cached Zebra state from `v25.0.0` to the latest disk format, delete the version file.
```

File: book/src/dev/state-db-upgrades.md [hunk 4]
```diff
@@ -62,17 +116,50 @@ But we need to use iterators for some operations, so our alternatives are (in pr
...
+To upgrade from a specific version `v25.x.y`, edit the file so it contains `x.y`.
+
+Editing the file will trigger a re-upgrade over an existing state. Re-upgrades can hide format bugs.
+For example, if the old code was correct, and the new code skips blocks, the validity checks won't
+find that bug.
```
Comment (arya2) [inline book/src/dev/state-db-upgrades.md]: Optional/Nitpick:

```suggestion
Editing the file and running Zebra will trigger a re-upgrade over an existing state.
Re-upgrades can hide format bugs. For example, if the old code was correct, and the
new code skips blocks, the validity checks won't find that bug.
```

File: book/src/dev/state-db-upgrades.md [hunk 5]
```diff
@@ -200,6 +296,7 @@ Each column family handles updates differently, based on its specific consensus
   - Each key-value entry is created once.
   - Keys can be deleted, but values are never updated.
   - Code called by ReadStateService must ignore deleted keys, or use a read lock.
+  - We avoid deleting keys, and avoid using iterators on Delete column families, for performance.
```
Comment (arya2) [inline book/src/dev/state-db-upgrades.md]:
```suggestion
  - We avoid deleting keys, and avoid using iterators on `Delete` column families, for performance.
```

**Review by arya2 — DISMISSED** [review]
Looks good!

File: book/src/dev/state-db-upgrades.md (hunk 2 above)
Comment (teor2345) [reply book/src/dev/state-db-upgrades.md]: Thank you for the detailed review and feedback!

This really good feedback, even though it is not quite right - that means we need to clearly document the current behaviour.

```suggestion
- Zebra must be able to upgrade the format from previous minor or patch versions of its disk format
  (Major disk format versions are breaking changes. They create a new empty state and re-sync the whole chain.)
```

**Review by upbqdn — APPROVED** [review]

File: book/src/dev/state-db-upgrades.md [hunk 6]
```diff
@@ -1,18 +1,64 @@
...
 For most state upgrades, we want to modify the database format of the existing database. If we
 change the major database version, every user needs to re-download and re-verify all the blocks,
 which can take days.
 
+### Writing Blocks to the State
+[write-block]: #write-block
+
+Blocks can be written to the database via two different code paths, and both must produce the same results:
+
+- Upgrading pre-existing zebra-state databases/cache to the latest format
```
Comment (upbqdn) [inline book/src/dev/state-db-upgrades.md]: ###### Optional Suggestion

Could we simplify this line to something like:

```suggestion
- Upgrading a pre-existing database to the latest format
```

**Review by upbqdn — DISMISSED** [review]

File: book/src/dev/state-db-upgrades.md (hunk 6 above)
Comment (teor2345) [reply book/src/dev/state-db-upgrades.md]: Sure!

**Comment by teor2345** [comment]
This has had two full reviews now.

**Review by upbqdn — APPROVED** [review]

**Pull request merged**